### PR TITLE
Implement Translatable Doctrine Extension

### PR DESCRIPTION
**Pull request by JarJak**
- refactor $this->fields to a name-indexed array to make field existence check much faster (simple isset instead of foreaching)
- rename get to getField, has to hasField
- implement Translatable instead of reinventing the wheel

First part of #151

### CODE REVIEW

File: src/Entity/Field.php [hunk 1]
```diff
@@ -190,7 +196,7 @@ public function getValue(): ?array
      *
      * @return array|mixed|null
      */
-    public function getFieldValue()
+    public function getFlatenValue()
```
Comment (bobdenotter) [inline src/Entity/Field.php]: Really like where this PR is going. But, this is a bit of an odd Method name.

Comment (JarJak) [reply src/Entity/Field.php]: What do you propose? It returns single value if count($this->getValues()) === 1, array if it's >1 and null if it is zero.

Comment (bobdenotter) [reply src/Entity/Field.php]: Then it should be `getFlatValue` or `getFlattenedValue`. "Flaten" looks like a typo. 😇

**Comment by bobdenotter** [comment]
Works like a charm for me! 

Could we change the name of the new table from `ext_translations` to perhaps `bolt_translations? I found some info about that here: https://github.com/Atlantic18/DoctrineExtensions/issues/1129

**Comment by bobdenotter** [comment]
To me, "a lot of queries" is no problem. "slow queries" are. ;-)

If O look at it, i see: 

schermafbeelding 2019-01-23 om 09 30 52

So, about 5% of resources are taken by doctrine now (4.3 ms / 86 ms). Of those 4.3 ms the majority happens in the "sibebar", which we'll be caching. So, i am fine with this solution. 

schermafbeelding_2019-01-23_om_09_30_39

I see `KnpLabs/DoctrineBehaviors` is "looking for maintainers", which doesn't say anything conclusive, but it's not a good sign that it's actively maintained. 

Finally, `Atlantic18/DoctrineExtensions` has `Tree`, which is something i would really, really love to have for Content, so we can make hierarchical content out of the box.

**Comment by JarJak** [comment]
Interesting... this is what I see:
zrzut ekranu z 2019-01-23 09-47-58

**Comment by bobdenotter** [comment]
Hmm, I'm not sure if I was looking at the correct branch earlier.. Just now i tried again, and i'm getting similar numbers as you.

Looks like it does a separate query for each field that is translated. In my case it's especially the sidebar that does a lot of queries. I wonder if this is "the price we pay" for baking in i18n, or if there's something we can do about it. 

I propose we keep it like this for now, then I add caching to the sidebar, and then it'll be easier to properly inspect which queries it does exactly. Is that a plan?

**Comment by JarJak** [comment]
@bobdenotter I've made db tables prefix fully configurable

File: src/Entity/ContentLocalizeTrait.php [hunk 2]
```diff
@@ -19,71 +19,68 @@ public function getLocales()
...
-
-        // Fourth: see if we can create the field on the fly
-        if (! empty($definition)) {
-            return Field::factory($definition, $name);
-        }
-
-        // Alas, return an empty generic Field
-        return Field::factory(['type' => 'generic']);
-    }
+//    public function getLocalizedFields(string $locale = '', bool $fallback = true): Collection
```
Comment (bobdenotter) [inline src/Entity/ContentLocalizeTrait.php]: If it's working correctly, shouldn't we just remove this chunk of commented-out code?

Comment (JarJak) [reply src/Entity/ContentLocalizeTrait.php]: removed

**Pull request merged**